### ISSUE TITLE
[WIP] RFC: Postgres support

### ISSUE THREAD

**Issue by deepthi** [vitessio/vitess#7084]
Before anyone gets too excited, this work is not yet planned by anyone we know. The intent behind writing this up is to document the size and scope of the work and a possible blueprint for when someone is ready to take this up. This RFC is intentionally at a very high-level. Each of the sections will probably need a more detailed RFC of their own.

## Feature Description
Vitess runs with MySQL as the backend DB. This proposal is to allow Vitess to run with PostgreSQL as the backend DB.

## Use Case(s)
* People who are already running on Postgres.
* Applications that would like to use Postgres features that are not supported by MySQL.

## Areas of Work
Adding Postgres will require work in all major parts of the Vitess codebase.

#### Vtgate
Vtgate currently speaks the mysql protocol. This means that clients can connect to vitess as if it is a plain mysql server. We will need to implement the equivalent server-side support so that Postgres clients can connect to vtgate. 

#### Parser
Vitess parser understands MySQL. PostgreSQL is not very different but there might be differences in syntax that we need to handle. There will be some high-level design decisions that have to be made here. For instance, does the vitess sql grammar become a superset of MySQL and PostgreSQL or do we have two grammars and two parsers?

#### TabletServer
TabletServer connects to the underlying DB and maintains connection pools. All of this will need to work with Postgres.

#### Mysqlctl
When running in managed mode, vitess uses mysqlctl to manage the mysql instances. We'll need an equivalent pgctl process that knows how to manage Postgres instances.

#### TabletManager
TabletManager provides an API that is used by various cluster management components. It will need to know whether the underlying DB is MySQL/Postgres in order to issue the necessary low-level commands.

#### Vtctld / wrangler / vtorc
Cluster management functions will probably not need to change as long as they are communicating to tablets through TabletManager. The exception is vtorc because it currently uses a mix of vttablet and mysql commands.

### VReplication
VReplication is closely tied to mysql's binlog replication. A significant portion of existing code will have to be re-written for Postgres.

## Project Phases
It should be obvious by now that adding Postgres support will require a large amount of work. However, it could be added in phases. One way of phasing it would be as follows.

### Phase 1 - Unmanaged, unsharded
To put vitess in front of an existing Postgres deployment, we might need only the Query serving portion of the work.

### Phase 2 - Managed, unsharded
Implementing the Cluster Management portion of the work will allow us to support a managed Postgres vitess cluster.

### Phase 3 - Managed, sharded
Once VReplication works with Postgres, we can support sharding. All other VReplication functionality (Materialize, MoveTables etc.) will also be available at that time.

**Comment by harshit-gangal** [comment]
I think there can be Phase 4 which can be re-sharding for which we require VReplication. In Phase 3 it can be without VReplication support but supporting sharded postgres.

**Comment by Frank591** [comment]
@deepthi, thanks for the inititative. It's really cool that you have started it. As I know Postgres support is very actual for some companies.  Do you have any time estimation for works which required to achieve each phase? And are there any known blockers that can prevent the implementation of Postgres support?

**Comment by deepthi** [comment]
I think in the past we have estimated this work at around 1 year for a team of 3-4 people. I don't know of any fundamental limitations on the vitess side that prevent this work.

**Comment by deepthi** [comment]
Update on 5/25/2021:
The Vitess maintainer team currently has no plans to add Postgres support to Vitess. If the Postgres community (or a Postgres user) would like to take this project up, the best way to proceed would be to do it as a fork of Vitess. Once the implementation has been proven in production, a “merge” project can be planned to bring the fork back into upstream.

**Comment by FruityWelsh** [comment]
On the Parser question, would a common grammar and subsets for implantation specific grammar make sense?

**Comment by FruityWelsh** [comment]
is [pg_ctl](https://www.postgresql.org/docs/10/app-pg-ctl.html) the equivalent to Mysqlctl?

**Comment by JensvandeWiel** [comment]
Currently working on a project where Vitess with a Postgres backend can be a great tool. Ill look into phase 1, Im new to the project but not new to Golang. 

Ill start looking into the working of Vitess and go ahead and try to implement phase 1, let me know you have any tips or good starting points.

**Comment by Bessonov** [comment]
FYI: there is [Citus](https://www.citusdata.com/) and it is doing something similar to Vitess.

**Comment by JensvandeWiel** [comment]
Yes, I know about Citus, but since I want to make a platform that supports multiple distributed databases. Because of this, it would be nice if Vitess would handle Mysql and Postgresql, I'll probably start with Citus or Yugabyte for now, since even if Postgres support is added it is still far away.

**Comment by harshit-gangal** [comment]
> Currently working on a project where Vitess with a Postgres backend can be a great tool. Ill look into phase 1, Im new to the project but not new to Golang.
> 
> Ill start looking into the working of Vitess and go ahead and try to implement phase 1, let me know you have any tips or good starting points.
> 
> Help would be lovely

The starting point would be to get the Postgres protocol implemented in Vitess.

**Comment by shayonj** [comment]
For my own learning - What would be the advantage of starting from scratching in vitess vs say starting a new project? Given there is work involved of implementing PG protocol from ground up. 

Excited for something in this space for PG nonetheless

**Comment by JensvandeWiel** [comment]
In my opinion it's better to keep it under one project because you can re-use code. PostgreSQL is different but there is still reusable code.

**Comment by Denchick** [comment]
> The starting point would be to get the Postgres protocol implemented in Vitess.

We were thinking about adding PostgreSQL support to Vitess several years ago. However, we realized that it would be more efficient if each tool was focused on a specific task.

You can take a look at [SPQR](http://github.com/pg-sharding/spqr/).

**Comment by JensvandeWiel** [comment]
After looking into the source code i've come to the same conclusion. Maybe its a good idea to mark this as not planned.
